File: Difficulty: Hard/Alien Dictionary/alien-dictionary.cpp
class Solution {
  public:
 
    string findOrder(vector<string> &words) {
        unordered_map<char, vector<char>> adj;
        unordered_map<char, int> indegree;
        unordered_set<char> unique_chars;

        // Initialize unique characters and indegree
        for (const string& word : words) {
            for (char c : word) {
                unique_chars.insert(c);
            }
        }
        for (char c : unique_chars) {
            indegree[c] = 0;
        }

        // Build the graph
        for (int i = 0; i < words.size() - 1; ++i) {
            const string& word1 = words[i];
            const string& word2 = words[i + 1];
            int len = min(word1.size(), word2.size());
            bool found = false;
            for (int j = 0; j < len; ++j) {
                char c1 = word1[j];
                char c2 = word2[j];
                if (c1 != c2) {
                    adj[c1].push_back(c2);
                    indegree[c2]++;
                    found = true;
                    break;
                }
            }
            // Handle cases like ["abc", "ab"] which is invalid
            if (!found && word1.size() > word2.size()) {
                return "";
            }
        }

        // Kahn's algorithm for topological sort
        queue<char> q;
        for (const auto& pair : indegree) {
            if (pair.second == 0) {
                q.push(pair.first);
            }
        }

        string result;
        while (!q.empty()) {
            char c = q.front();
            q.pop();
            result += c;
            for (char neighbor : adj[c]) {
                indegree[neighbor]--;
                if (indegree[neighbor] == 0) {
                    q.push(neighbor);
                }
            }
        }

        // Check if all characters were included (no cycle)
        if (result.size() != unique_chars.size()) {
            return "";
        }

        return result;
    } 
};

File: Difficulty: Hard/Alien Dictionary/alien-dictionary.py
from collections import deque

class Solution:
    def findOrder(self, words):
        # Create adjacency list for 26 letters
        adj = [[] for _ in range(26)]
        
        # Build graph
        for i in range(len(words) - 1):
            s1 = words[i]
            s2 = words[i + 1]
            length = min(len(s1), len(s2))
            found_diff = False
            for ptr in range(length):
                if s1[ptr] != s2[ptr]:
                    u = ord(s1[ptr]) - ord('a')
                    v = ord(s2[ptr]) - ord('a')
                    adj[u].append(v)
                    found_diff = True
                    break
            # If no difference found and s1 is longer than s2, invalid order
            if not found_diff and len(s1) > len(s2):
                return ""
        
        # Find all unique characters used
        present = [False] * 26
        for word in words:
            for ch in word:
                present[ord(ch) - ord('a')] = True
        unique_chars = sum(present)
        
        # Build indegree array
        indegree = [0] * 26
        for i in range(26):
            for it in adj[i]:
                indegree[it] += 1
        
        # Topological sort BFS
        q = deque()
        for i in range(26):
            if present[i] and indegree[i] == 0:
                q.append(i)
        
        ans = []
        while q:
            node = q.popleft()
            ans.append(node)
            for it in adj[node]:
                indegree[it] -= 1
                if indegree[it] == 0:
                    q.append(it)
        
        # Check if all unique chars are in result (no cycle)
        if len(ans) != unique_chars:
            return ""
        
        # Convert back to characters
        return ''.join(chr(ord('a') + idx) for idx in ans)